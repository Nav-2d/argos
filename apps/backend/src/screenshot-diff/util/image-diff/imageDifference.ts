/* eslint-disable import/namespace */
import { compare } from "odiff-bin";

import { tmpName } from "@/storage/index.js";
import type { ImageFile } from "@/storage/index.js";

const createDifference = async (options: {
  baseImageFilepath: string;
  compareImageFilepath: string;
  diffImageFilepath: string;
}) => {
  const result = await compare(
    options.baseImageFilepath,
    options.compareImageFilepath,
    options.diffImageFilepath,
    {
      outputDiffMask: true,
      threshold: 0.061,
    },
  );

  if (result.match) {
    return 0;
  }

  switch (result.reason) {
    case "file-not-exists":
      throw new Error(`File not exists`);
    case "layout-diff":
      return 1;
    case "pixel-diff":
      return result.diffPercentage / 100;
    default:
      throw new Error(`Unknown reason`);
  }
};

async function getMaxDimensions(images: ImageFile[]) {
  const imagesDimensions = await Promise.all(
    images.map(async (image) => image.getDimensions()),
  );

  return {
    width: Math.max(...imagesDimensions.map(({ width }) => width)),
    height: Math.max(...imagesDimensions.map(({ height }) => height)),
  };
}

export default async function imageDifference(optionsWithoutDefault: {
  baseImage: ImageFile;
  compareImage: ImageFile;
}) {
  const { baseImage, compareImage } = optionsWithoutDefault;

  const [maxDimensions, diffImageFilepath] = await Promise.all([
    getMaxDimensions([baseImage, compareImage]),
    tmpName({ postfix: ".png" }),
  ]);

  // Resize images to the maximum dimensions
  const [baseImageFilepath, compareImageFilepath] = await Promise.all([
    baseImage.enlarge(maxDimensions),
    compareImage.enlarge(maxDimensions),
  ]);

  // Create difference
  const score = await createDifference({
    baseImageFilepath,
    compareImageFilepath,
    diffImageFilepath,
  });

  return {
    ...maxDimensions,
    filepath: diffImageFilepath,
    value: score,
  };
}
