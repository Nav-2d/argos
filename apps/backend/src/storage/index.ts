export * from "./checkIfExists.js";
export * from "./download.js";
export * from "./get.js";
export * from "./getPublicUrl.js";
export * from "./upload.js";
export * from "./signed-url.js";
export * from "./ImageFile.js";
export * from "./s3.js";
