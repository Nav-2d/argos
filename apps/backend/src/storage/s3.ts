import { S3Client } from "@aws-sdk/client-s3";

let client: S3Client;

export type { S3Client } from "@aws-sdk/client-s3";

export const getS3Client = () => {
  if (!client) {
    client = new S3Client({ region: "eu-west-1" });
  }
  return client;
};
