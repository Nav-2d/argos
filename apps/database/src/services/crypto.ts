import { randomBytes } from "node:crypto";
import { promisify } from "node:util";

const generateRandomBytes = promisify(randomBytes);

export const generateRandomHexString = async () => {
  const token = await generateRandomBytes(20);
  return token.toString("hex");
};
