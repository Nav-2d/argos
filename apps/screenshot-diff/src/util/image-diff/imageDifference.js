/* eslint-disable import/namespace */
import { stat, mkdir, unlink } from "node:fs/promises";
import { dirname } from "node:path";
import { spawn } from "node:child_process";
import { promisify } from "node:util";
import gm from "gm";
import { tmpName as cbTmpName } from "tmp";

const tmpName = promisify(cbTmpName);

const gmMagick = gm.subClass({ imageMagick: true });

const checkFileExists = async (filename) => {
  try {
    await stat(filename);
    return true;
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }
    throw error;
  }
};

const transparent = (filename, { width, height }) => {
  const gmImage = gmMagick(filename);
  gmImage.background("transparent"); // Fill in new space with white background
  gmImage.gravity("NorthWest"); // Anchor image to upper-left
  gmImage.extent(width, height); // Specify new image size
  return gmImage;
};

const getImageSize = async (filename) => {
  const gf = gmMagick(filename);
  return promisify(gf.size.bind(gf))();
};

export const handleRaw = (difference) => {
  const { raw, ...other } = difference;

  const matches = raw.match(/all: (.+)\n/);
  if (!matches) {
    throw new Error(`Expected raw to contain 'all' but received "${raw}"`);
  }
  return {
    ...other,
    value: parseFloat(matches[1], 10),
  };
};

const resizeImage = async (filename, { width, height }) => {
  const tmpFilename = await tmpName({ postfix: ".png" });
  const gf = transparent(filename, { width, height });
  await promisify(gf.write.bind(gf))(tmpFilename);
  return tmpFilename;
};

const getDiffArgs = ({
  actualFilename,
  expectedFilename,
  diffFilename,
  highlightColor,
  lowlightColor,
  fuzz,
}) => {
  const diffArgs = [
    "-verbose",
    "-highlight-color",
    highlightColor,
    "-lowlight-color",
    lowlightColor,
    // http://legacy.imagemagick.org/script/command-line-options.php#metric
    // http://www.imagemagick.org/Usage/compare/
    // https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/compare.c
    "-metric",
    "AE",
  ];

  if (fuzz) {
    diffArgs.push("-fuzz", fuzz);
  }

  diffArgs.push(
    actualFilename,
    expectedFilename,
    // If there is no output image, then output to `stdout` (which is ignored)
    diffFilename || "-"
  );
  return diffArgs;
};

const createDifference = async (options) => {
  const diffArgs = getDiffArgs(options);
  return new Promise((resolve, reject) => {
    // http://www.imagemagick.org/script/compare.php
    const proc = spawn("compare", diffArgs);
    let stderr = "";
    proc.stderr.on("data", (data) => {
      stderr += data;
    });
    proc.on("close", (code) => {
      // ImageMagick returns err code 2 if err, 0 if similar, 1 if dissimilar
      if (code === 0 || code === 1) {
        resolve(stderr);
        return;
      }
      reject(stderr);
    });
  });
};

const resizeIfNecessary = async (filename, dimensions, maxDimensions) => {
  if (
    dimensions.width !== maxDimensions.width ||
    dimensions.height !== maxDimensions.height
  ) {
    return resizeImage(filename, {
      width: maxDimensions.width,
      height: maxDimensions.height,
    });
  }
  return filename;
};

export async function rawDifference(options) {
  const { actualFilename, expectedFilename, diffFilename, ...other } = options;

  // Check if files exists
  await Promise.all(
    [actualFilename, expectedFilename].map(async (filename) => {
      if (!(await checkFileExists(filename))) {
        throw new Error(`File "${filename}" does not exist`);
      }
    })
  );

  const [actualSize, expectedSize] = await Promise.all([
    getImageSize(actualFilename),
    getImageSize(expectedFilename),
    diffFilename ? mkdir(dirname(diffFilename), { recursive: true }) : null,
  ]);

  // Find the maximum dimensions
  const maxDimensions = {
    width: Math.max(actualSize.width, expectedSize.width),
    height: Math.max(actualSize.height, expectedSize.height),
  };

  // Resize images to the maximum dimensions
  const [resizedActualFilename, resizedExpectedFilename] = await Promise.all([
    resizeIfNecessary(actualFilename, actualSize, maxDimensions),
    resizeIfNecessary(expectedFilename, expectedSize, maxDimensions),
  ]);

  // Create difference
  const raw = await createDifference({
    ...other,
    actualFilename: resizedActualFilename,
    expectedFilename: resizedExpectedFilename,
    diffFilename,
  });

  // Remove temporary files
  await Promise.all([
    resizedActualFilename !== actualFilename
      ? unlink(resizedActualFilename)
      : null,
    resizedExpectedFilename !== expectedFilename
      ? unlink(resizedExpectedFilename)
      : null,
  ]);

  return { ...maxDimensions, raw };
}

export default async function imageDifference(optionsWithoutDefault) {
  const {
    actualFilename,
    expectedFilename,
    highlightColor = "red",
    lowlightColor = "white",
    fuzz = 0,
    ...other
  } = optionsWithoutDefault;

  // Assert our options are passed in
  if (!actualFilename) {
    throw new Error(
      "`options.actualFilename` was not passed to `image-difference`"
    );
  }

  if (!expectedFilename) {
    throw new Error(
      "`options.expectedFilename` was not passed to `image-difference`"
    );
  }

  const options = {
    actualFilename,
    expectedFilename,
    highlightColor,
    lowlightColor,
    fuzz,
    ...other,
  };

  const difference = await rawDifference(options);
  return handleRaw(difference, options);
}
