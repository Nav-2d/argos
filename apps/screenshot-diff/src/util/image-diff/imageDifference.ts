/* eslint-disable import/namespace */
import { spawn } from "child_process";
import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";

import type { S3Image } from "@argos-ci/storage";

export const handleRaw = (diff: {
  raw: string;
  width: number;
  height: number;
}) => {
  const matches = diff.raw.match(/all: (.+)\n/);
  if (!matches) {
    throw new Error(`Expected raw to contain 'all' but received "${diff.raw}"`);
  }
  return {
    width: diff.width,
    height: diff.height,
    value: parseFloat(matches[1] as string),
  };
};

const getDiffArgs = ({
  baseImageFilePath,
  compareImageFilePath,
  diffImageFilename,
  highlightColor,
  lowlightColor,
  fuzz,
}: {
  baseImageFilePath: string;
  compareImageFilePath: string;
  diffImageFilename: string;
  highlightColor: string;
  lowlightColor: string;
  fuzz: string | number;
}) => {
  const diffArgs = [
    "-verbose",
    "-highlight-color",
    highlightColor,
    "-lowlight-color",
    lowlightColor,
    "-compose",
    "src",
    // http://legacy.imagemagick.org/script/command-line-options.php#metric
    // http://www.imagemagick.org/Usage/compare/
    // https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/compare.c
    "-metric",
    "AE",
  ];

  if (fuzz) {
    diffArgs.push("-fuzz", String(fuzz));
  }

  diffArgs.push(
    baseImageFilePath,
    compareImageFilePath,
    // If there is no output image, then output to `stdout` (which is ignored)
    diffImageFilename || "-"
  );
  return diffArgs;
};

const createDifference = async (options: {
  baseImageFilePath: string;
  compareImageFilePath: string;
  diffImageFilename: string;
  highlightColor: string;
  lowlightColor: string;
  fuzz: string | number;
}) => {
  const diffArgs = getDiffArgs(options);

  return new Promise<string>((resolve, reject) => {
    // http://www.imagemagick.org/script/compare.php
    const proc = spawn("compare", diffArgs);
    let stderr = "";
    proc.stderr.on("data", (data) => {
      stderr += data;
    });
    proc.on("close", (code) => {
      // ImageMagick returns err code 2 if err, 0 if similar, 1 if dissimilar
      if (code === 0 || code === 1) {
        resolve(stderr);
        return;
      }
      reject(stderr);
    });
  });
};

async function getMaxDimensions(images: S3Image[]) {
  const imagesDimensions = await Promise.all(
    images.map(async (image) => {
      const { width, height } = await image.getDimensions();
      if (!width || !height) await image.measureDimensions();
      return image.getDimensions();
    })
  );

  return {
    width: Math.max(
      ...(imagesDimensions.map(({ width }) => width).filter(Number) as [number])
    ),
    height: Math.max(
      ...(imagesDimensions.map(({ height }) => height).filter(Number) as [
        number
      ])
    ),
  };
}

export default async function imageDifference(optionsWithoutDefault: {
  baseImage: S3Image;
  compareImage: S3Image;
  diffImage: S3Image;
  highlightColor?: string;
  lowlightColor?: string;
  fuzz?: string | number;
}) {
  const {
    baseImage,
    compareImage,
    diffImage,
    highlightColor = "red",
    lowlightColor = "none",
    fuzz = "0",
  } = optionsWithoutDefault;

  const [maxDimensions] = await Promise.all([
    getMaxDimensions([baseImage, compareImage]),
    mkdir(dirname(diffImage.getTargetFilePath()), { recursive: true }),
  ]);

  // Resize images to the maximum dimensions
  await Promise.all([
    baseImage.enlarge(maxDimensions),
    compareImage.enlarge(maxDimensions),
  ]);

  // Create difference
  const raw = await createDifference({
    highlightColor,
    lowlightColor,
    fuzz,
    baseImageFilePath: baseImage.getFilePath(),
    compareImageFilePath: compareImage.getFilePath(),
    diffImageFilename: diffImage.getTargetFilePath(),
  });
  diffImage.setLocalFile(true);

  return handleRaw({ ...maxDimensions, raw });
}
