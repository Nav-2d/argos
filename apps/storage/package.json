{
  "name": "@argos-ci/storage",
  "version": "2.0.0",
  "private": true,
  "exports": "./dist/index.js",
  "type": "module",
  "dependencies": {
    "@argos-ci/config": "*",
    "@aws-sdk/client-s3": "^3.386.0",
    "@aws-sdk/s3-request-presigner": "^3.386.0",
    "mime": "^3.0.0",
    "sharp": "^0.32.4",
    "tmp": "^0.2.1"
  },
  "devDependencies": {
    "@types/mime": "^3.0.1",
    "@types/sharp": "^0.31.1",
    "@types/tmp": "^0.2.3",
    "axios": "^1.4.0"
  },
  "scripts": {
    "build": "swc src --config-file ../../.swcrc -d dist",
    "watch-build": "npm run build -- --watch --quiet"
  }
}
