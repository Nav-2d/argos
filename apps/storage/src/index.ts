export * from "./checkIfExists.js";
export * from "./download.js";
export * from "./get.js";
export * from "./upload.js";
export * from "./signed-url.js";
export * from "./s3.js";
