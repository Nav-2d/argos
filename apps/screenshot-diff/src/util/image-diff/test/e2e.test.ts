/* eslint-disable @typescript-eslint/no-unused-vars */
// @ts-ignore
import { createHash } from "node:crypto";
import { createReadStream } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
// @ts-ignore
import rimraf from "rimraf";

import { LocalImageFile } from "@argos-ci/storage";

import imageDifference from "../imageDifference.js";

const __dirname = fileURLToPath(new URL(".", import.meta.url));

export const hashFile = async (filepath: string) => {
  const fileStream = createReadStream(filepath);
  const hash = createHash("sha256");
  await new Promise((resolve, reject) => {
    fileStream.on("error", reject);
    hash.on("error", reject);
    hash.on("finish", resolve);
    fileStream.pipe(hash);
  });
  return hash.read().toString("hex");
};

async function compareLocalImages(params: {
  baseFilepath: string;
  compareFilepath: string;
}) {
  return imageDifference({
    baseImage: new LocalImageFile({
      filepath: params.baseFilepath,
    }),
    compareImage: new LocalImageFile({
      filepath: params.compareFilepath,
    }),
  });
}

describe("e2e", () => {
  beforeAll(async () => {
    // Clean up actual-files
    await rimraf.sync(join(__dirname, "/actual-files"));
  });

  it("generates different images", async () => {
    const baseFilename = "checkerboard.png";
    const compareFilename = "white.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });

  it("generates the same image", async () => {
    const baseFilename = "checkerboard.png";
    const compareFilename = "checkerboard.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();

    expect(difference).toEqual({
      value: 0,
      width: 10,
      height: 10,
    });
  });

  it("generates different sizes images", async () => {
    const baseFilename = "checkerboard-excess.png";
    const compareFilename = "checkerboard.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });

  it("generates the same image where 1 has a transparent background", async () => {
    const baseFilename = "checkerboard-transparent.png";
    const compareFilename = "checkerboard.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });

  it("generates images with big diff", async () => {
    const baseFilename = "old-site.png";
    const compareFilename = "new-site.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });

  it("works with large size image", async () => {
    const baseFilename = "big-image.png";
    const compareFilename = "big-image2.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });

  it("generates images similar colors", async () => {
    const baseFilename = "violet-square.png";
    const compareFilename = "black-square.png";

    const { filepath: diffFilepath, ...difference } = await compareLocalImages({
      baseFilepath: join(__dirname, "/test-files", baseFilename),
      compareFilepath: join(__dirname, "/test-files", compareFilename),
    });

    expect(difference).toMatchSnapshot();
  });
});
