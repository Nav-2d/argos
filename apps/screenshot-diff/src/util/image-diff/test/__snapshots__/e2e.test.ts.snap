// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e generates different images 1`] = `
{
  "height": 10,
  "value": 45,
  "width": 10,
}
`;

exports[`e2e generates different sizes images 1`] = `
{
  "height": 15,
  "value": 55.5555555556,
  "width": 15,
}
`;

exports[`e2e generates images with big diff 1`] = `
{
  "height": 4469,
  "value": 97.640956341,
  "width": 1000,
}
`;

exports[`e2e generates the same image 1`] = `
{
  "height": 10,
  "value": 0,
  "width": 10,
}
`;

exports[`e2e generates the same image where 1 has a transparent background 1`] = `
{
  "height": 10,
  "value": 50,
  "width": 10,
}
`;

exports[`e2e works with large size image 1`] = `
{
  "height": 13358,
  "value": 17.1186113661,
  "width": 900,
}
`;
